### DISCUSSION TITLE
How will users access ArchiveFiles like mzXML?

### DISCUSSION THREAD

**Discussion by mneinast** [Princeton-LSI-ResearchComputing/tracebase#1155] [Ideas]
The database now includes mzXML (and even other formats like .raw?), but we need to design how users interact with them.  Right now, users can find and download individual files using the ArchiveFile page, but this is really a placeholder / testing page.

At the highest level, there is 1) accessing the mzXML and 2) uploading new curated peaks for the mzXML which they have accessed.  
Of course, how the users actually analyze the mzXML is outside the scope of TraceBase.

## 1) accessing mzXML
What might users do with mzXML in TraceBase?  Essentially, they will look for PeakGroups not already included in the PeakAnnotation files.  I'm calling these "uncurated PeakGroups"...they may be:
 - PeakGroups whose chemical identity is not confirmed
    - maybe we should not accept these, although we currently don't have a way to enforce true identity anyway (we trust the user).
 - PeakGroups with a confirmed identification (ie glucose) that were not included when the data was submitted to tracebase

Here are some more specific examples:
 - Look for labeling in an uncurated PeakGroup in a samples from mice infused with a specific tracer
-> download a highly filtered subset of mzXML.
 - Look for labeling from ANY tracer in an uncurated PeakGroup a specific tissue (ie Brain)
-> download all mzXML associated with Brain samples.
 - Look for labeling in an uncurated PeakGroup from ANY tracer in ANY tissue.
-> download all mzXML?
 - Execute an automated peak-picking script on any set of mzXML, or perhaps all mzXML (someday)
-> download all mzXML?  Access via API?

So there are two basic "levels":  download a filtered subset of mzXML versus download all mzXML.  Maybe this is implemented with just a different version of the Advanced Search tool:  it would not show any data (because we're not always directly associating data with a single mzXML), but it would show the metadata for samples and indicate raw file availability.  Then maybe you have an option to download a zip of this filtered table + mzXML.

For accessing all mzXML, could we just give this person access to the tracebase directory, along with a table which identifies samples by location in msdata?  I'm imagining an excel sheet (or tsv) where each row is one mzXML file and the columns provide whatever identifying information we have.

Along those lines, if we're comfortable giving people access to the tracebase directory, could we just do this for filtered subsets of samples?  Give them a much smaller table with only the samples they're looking for?  Then there is no option to download a file on the site, but it's at least simple.

## 2) uploading newly curated peaks for mzXML downloaded from TraceBase

Assuming TraceBase always maintains its own copy of every mzXML, we could append a unique identifier to each mzXML filename.  For example, a 6-8 character code from a base-62 system (we'd have to actually estimate the number of files that TraceBase could track).  Maybe it's possible to also inject this into an mzXML without corrupting it?

Anyway, if Maven (or similar) is used to pull these peaks, then that unique identifier would be present in the sample name.  So when the user uploads this data, the sample name should match an existing sample name in the database.  If it helps, we could ask the user to indicate when they are uploading peaks from mzXML already in tracebase.  In this case, we already have all the sample information, so the user does not need to upload any study doc (and we don't even have to regenerate it).

**Comment by hepcat72** [comment]
A few thoughts...

## 1) accessing mzXML

Whether it's the advanced search interface or any of the other tables embedded in various pages, we could have a separate mzXML download button that creates (and initiates a download of) a zip file of the mzxml files in the results.  This means that we would have 2 download buttons: one for the table and one for mzXML files.

This would be quite doable, I think, given my recent experience replacing the existing download button with a custom one on the compounds list page.  (I'm even thinking of re-working the table toolbar on the advanced search page to better integrate the download and stats buttons to look like a single blue bar, which would appear more congruous.)

And given the fine-grained control of the advanced search, we wouldn't need any other form of compiling a list of mzXML files.

We would just have to add an mzXML column to the various results formats. The one issue there is that all of the available result formats of the advanced search would only ever display mzXML files that are associated with annotated peaks.  It would exclude unanalyzed mzXML records.  So what we would need is a results format that could include those.

Right now, every `mzXML` file requires an `MSRunSequence` record and a `Sample` record.  So we could probably base a new results format on the `Sample` table.  And users could search for samples whose `mzXML` file field "has a value" (to exclude records that don't have mzXMLs associated.  We could also have a "measured compound" column that users could use to only look for Samples that have no measured compounds (i.e. no annotated peak groups).  **Then** they could narrow those results based on polarity and scan range.  Perhaps we could also add retention time (range) as a field to `MSRunSample`.

## 2) uploading newly curated peaks for mzXML downloaded from TraceBase

We don't actually need curated peaks to be able to load `mzXML` files.  We don't even need an existing `Sample` or `MSRunSequence` record.  It will load `ArchiveFile` records without any `MSRunSample` links to the file.  At any time (before during or after `Sample` and/or `PeakGroup` loading, we can associate those `mzXML` files with data in a `Study`.  All we need is a sequence and an existing sample.  We have a script named `load_msruns` that has an `--mzxml-files` option (which we want to be able to supplement or replace with an `--mzxml-dir` option that could walk subdirectories and load all files).  You can run it with or without the `PeakAnnotation Details` sheet. If the file names match existing samples (with or without scan labels like `_pos` that, as you noticed, as automatically removed) and sufficient sequence details are provided to uniquely identify the sequence using the options `--researcher`, `--instrument`, `--lcprotocol`, and `--date`, it will associate the `mzXML` files with `MSRunSample` records.

So there's no need for a code embedded in the file to identify a sample.  If the file name doesn't match the sample name (accounting for scan labels), it can be associated manually by entering it into that `Peak AnnotationDetails` sheet.  The sample record just has to exist (in the DB already, or in the `Samples` sheet) before you can associate an `mzXML` with it.

The `load_msruns` script can load `mzXML` files before, during, or after the load of peak annotations.  I programmed it to be smart enough to associate peak annotations with the `MSRunSample` records containing the mzXML files regardless of when they're loaded.  So if the peak annotations were previously loaded, and we want to add `mzXML` files, when we run that loader, it will re-link those peak annotations: remove the link to the placeholder `MSRunSample` record that has to `mzXML` to a newly created one **with** the uploaded `mzXML` file associated.  If we want to associate previously loaded `mzXML` files with newly entered samples, we can run `load_msruns` with the sheets (and/or the files again).

The only problem it currently has in this regard is, if there are multiple `mzXML` files with the exact same name (i.e. no scan labels), it has no way to link the peak groups to the right one.  However, we have a plan to link them to **all** matching `MSRunSample` records with `mzXML` files.  That's described in issue #1045.

So I'm guessing that what you describe as "append a unique identifier to each mzXML filename" is effectively the mechanism we currently have with "scan labels" like `_pos`.  However, that's not going to scale well with automatic peak detection algorithms.  The association we're trying to distinguish is not just `mzXML` with `Sample`.  We can do that automatically, but it's the association of `mzXML` with `PeakGroup`/measured-`Compound`.  (We may not need this at all however - see my questions below.)

The `Peak Annotation Details` sheet provides a mechanism to make the association between `mzXML` and `Sample`.  The problem with it is that it would require laborious manual entry (by entering path data to distinguish between `mzXML`s of the same name).  The solution might be the `mzrolldb` file.  In that meeting we had with the maven folks, they indicated that the source file of the peak data is recorded in there.

Having users manually deal with the association (whether it is entry into the sheet, file renaming, or embedding data in the file), it's laborious either way.

So my question is:
- Do we **want** to be able to associate `PeakGroup`s with specific `mzXML` files?  We can do `mzXML` with `Sample` automatically.  Is that sufficient?  My understanding is that that's all we need.  But does that change with automatic peak detection?

**Reply by mneinast** [reply]
responding to Rob's thoughts:

## 1) accessing mzXML

I think we have the same vision.  Basically, use the Advanced Search to filter the samples, then download mzXML.  I like your ideas to handle the mzXML which do not have directly associated peaks.  they might not be necessary, though.  We could just always provide every associated mzXML (potentially with the filtering you describe for polarity, etc).

## 2) uploading...

It's cool that you don't need curated peaks to load mzXML - I will discuss with you in person to understand how this works.  It's definitely a huge challenge to link a PeakGroup to the correct mzXML given the current workflow and design of tracebase.  The mzrolldb files could, in theory, help.  However, I don't think it's perfect because it relies on relative filepaths.  I think that, in general, this is probably something for a version 4.0.

> Do we want to be able to associate PeakGroups with specific mzXML files? We can do mzXML with Sample automatically. Is that sufficient? My understanding is that that's all we need. But does that change with automatic peak detection?

No, we do not need to associate PeakGroups with specific mzXML.  I think we will be very happy if we just associate mzXML with Sample.

**Comment by mneinast** [comment]
Here are some mockups for how the site should provide access to mzXML.  https://docs.google.com/presentation/d/1_4chMS9HPnhFIxfvdTiyq3DVJAuAXVxR4bFi8bhjI0s/edit?usp=sharing
